### ISSUE TITLE
Testcase #1 in `distribution_derive.yaml` panics after inserting data

### ISSUE THREAD

**Issue by kwannoel** [risingwavelabs/risingwave#9843]
### Describe the bug

I ran this set of queries: https://github.com/risingwavelabs/risingwave/blob/e6080f290819d2a3cb512fc25323b27174d4f7ac/src/frontend/planner_test/tests/testdata/distribution_derive.yaml#L10
and inserted data into the upstream tables. See script below on how to reproduce.

`Failed to append datum, array builder type: Serial, scalar type: Int32`

<details closed>
<summary>compute logs </summary>

```
thread 'risingwave-streaming-actor' panicked at 'Failed to append datum, array builder type: Serial, scalar type: Int32', src/common/src/array/mod.rs:570:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:577:5
   1: core::panicking::panic_fmt
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/panicking.rs:67:14
   2: risingwave_common::array::ArrayBuilderImpl::append_datum_n
             at ./src/common/src/array/mod.rs:519:60
   3: risingwave_common::array::ArrayBuilderImpl::append_datum
             at ./src/common/src/array/mod.rs:530:17
   4: risingwave_stream::common::builder::StreamChunkBuilder::append_row
             at ./src/stream/src/common/builder.rs:141:13
   5: risingwave_stream::executor::lookup::impl_::<impl risingwave_stream::executor::lookup::LookupExecutor<S>>::execute_inner::{{closure}}
             at ./src/stream/src/executor/lookup/impl_.rs:318:50
   6: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
   7: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
   8: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
   9: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:1632:9
  10: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/next.rs:32:9
  11: <await_tree::future::Instrumented<F,_> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/await-tree-0.1.1/src/future.rs:124:23
  12: risingwave_stream::executor::wrapper::trace::instrument_await_tree::{{closure}}
             at ./src/stream/src/executor/wrapper/trace.rs:124:10
  13: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  14: risingwave_stream::executor::wrapper::schema_check::schema_check::{{closure}}
             at ./src/stream/src/executor/wrapper/schema_check.rs:24:1
  15: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  16: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  17: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:1632:9
  18: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/next.rs:32:9
  19: risingwave_stream::executor::wrapper::epoch_check::epoch_check::{{closure}}
             at ./src/stream/src/executor/wrapper/epoch_check.rs:31:44
  20: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  21: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  22: <S as futures_core::stream::TryStream>::try_poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:196:9
  23: futures_util::stream::try_stream::TryStreamExt::try_poll_next_unpin
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/try_stream/mod.rs:1068:9
  24: <futures_util::stream::try_stream::try_next::TryNext<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/try_stream/try_next.rs:32:9
  25: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:348:35
  26: tokio::task::task_local::LocalKey<T>::scope_inner
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:233:19
  27: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:345:13
  28: risingwave_common::util::epoch::task_local::scope::{{closure}}
             at ./src/common/src/util/epoch.rs:176:46
  29: risingwave_stream::executor::wrapper::epoch_provide::epoch_provide::{{closure}}
             at ./src/stream/src/executor/wrapper/epoch_provide.rs:31:59
  30: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  31: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  32: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:1632:9
  33: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/next.rs:32:9
  34: <minitrace::future::InSpan<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/minitrace-0.4.1/src/future.rs:118:19
  35: risingwave_stream::executor::wrapper::trace::trace::{{closure}}
             at ./src/stream/src/executor/wrapper/trace.rs:53:60
  36: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  37: risingwave_stream::executor::wrapper::update_check::update_check::{{closure}}
             at ./src/stream/src/executor/wrapper/update_check.rs:27:1
  38: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  39: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  40: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  41: <risingwave_stream::executor::dispatch::DispatchExecutor as risingwave_stream::executor::StreamConsumer>::execute::{{closure}}
             at ./src/stream/src/executor/dispatch.rs:257:9
  42: <futures_async_stream::try_stream::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-async-stream-0.2.6/src/lib.rs:463:33
  43: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:120:9
  44: <&mut S as futures_core::stream::Stream>::poll_next
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/stream.rs:104:9
  45: <tokio_stream::stream_ext::next::Next<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/git/checkouts/tokio-968c02b7a1a41bea/ab251ad/tokio-stream/src/stream_ext/next.rs:42:9
  46: <tokio_stream::stream_ext::try_next::TryNext<St> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/git/checkouts/tokio-968c02b7a1a41bea/ab251ad/tokio-stream/src/stream_ext/try_next.rs:43:9
  47: <minitrace::future::InSpan<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/minitrace-0.4.1/src/future.rs:118:19
  48: <await_tree::future::Instrumented<F,_> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/await-tree-0.1.1/src/future.rs:124:23
  49: risingwave_stream::executor::actor::Actor<C>::run_consumer::{{closure}}
             at ./src/stream/src/executor/actor.rs:186:17
  50: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/future/maybe_done.rs:68:48
  51: risingwave_stream::executor::actor::Actor<C>::run::{{closure}}::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/macros/join.rs:126:24
  52: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/future/poll_fn.rs:58:9
  53: risingwave_stream::executor::actor::Actor<C>::run::{{closure}}
             at ./src/stream/src/executor/actor.rs:154:9
  54: risingwave_stream::task::stream_manager::LocalStreamManagerCore::build_actors::{{closure}}::{{closure}}
             at ./src/stream/src/task/stream_manager.rs:648:50
  55: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:348:35
  56: tokio::task::task_local::LocalKey<T>::scope_inner
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:233:19
  57: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:345:13
  58: await_tree::registry::TreeRoot::instrument::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/await-tree-0.1.1/src/registry.rs:51:44
  59: <futures_util::future::either::Either<A,B> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/future/either.rs:109:32
  60: <tokio_metrics::task::Instrumented<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-metrics-0.1.0/src/task.rs:2365:19
  61: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  62: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:223:17
  63: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/loom/std/unsafe_cell.rs:14:9
  64: tokio::runtime::task::core::Core<T,S>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:212:13
  65: tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:476:19
  66: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/panic/unwind_safe.rs:271:9
  67: std::panicking::try::do_call
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:485:40
  68: ___rust_try
  69: std::panicking::try
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:449:19
  70: std::panic::catch_unwind
             at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panic.rs:140:14
  71: tokio::runtime::task::harness::poll_future
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:464:18
  72: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:198:27
  73: tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:152:15
  74: tokio::runtime::task::raw::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:255:5
  75: tokio::runtime::task::raw::RawTask::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:200:18
  76: tokio::runtime::task::LocalNotified<S>::run
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/mod.rs:394:9
  77: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:487:21
  78: tokio::runtime::coop::with_budget
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:107:5
  79: tokio::runtime::coop::budget
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:73:5
  80: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:463:9
  81: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:433:24
  82: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:406:17
  83: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/macros/scoped_tls.rs:61:9
  84: tokio::runtime::scheduler::multi_thread::worker::run
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:403:5
  85: tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:365:45
  86: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/blocking/task.rs:42:21
  87: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  88: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:223:17
  89: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/loom/std/unsafe_cell.rs:14:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


*** await tree context of current task ***

Actor 50: `CREATE MATERIALIZED VIEW m1 AS SELECT A.v, B.v AS Bv FROM A JOIN B USING(k1)` [1.652s]
  Epoch 4390623177932800 [1.644s]
    LookupExecutor (actor 50, executor 4294967288) [1.644s]  <== current
      MergeExecutor 32FFFFFFFC (actor 50, executor 4294967292) [1.640s]


Tue May 16 09:53:14 UTC 2023 [risedev]: Program exited with 134

```

</details>

### To Reproduce

<details closed>
<summary>script</summary>

```
#!/usr/bin/env bash
./risedev k
./risedev clean-data
./risedev d
psql -h localhost -p 4566 -d dev -U root -c "SET RW_IMPLICIT_FLUSH=true;
set rw_streaming_enable_delta_join = true;
create table A      (k1 int, k2 int, k3 int, v int);
create index Ak1   on A(k1) include(k1,k2,k3,v);
create index Ak1k2 on A(k1,k2) include(k1,k2,k3,v);
create table B      (k1 int, k2 int, k3 int, v int);
create index Bk1   on B(k1) include(k1,k2,k3,v);
insert into A values(1, 2, 3, 4);
insert into B values(1, 2, 3, 4);
CREATE MATERIALIZED VIEW m1 as select A.v, B.v as Bv from A join B using(k1);
select * from m1;
"
./risedev k
```

</details>

### Expected behavior

No panic

**Comment by fuyufjh** [comment]
@yufansong Any updates?

**Comment by yufansong** [comment]
> @yufansong Any updates?

Not yet. Busy with some other work. If needed, I can focus on this issue next week.

**Comment by yufansong** [comment]
Get it, will work on it this week.

**Comment by stdrc** [comment]
This looks very similar to https://github.com/risingwavelabs/risingwave/issues/12620. Cc @yufansong can refer to that for some inspiration. The direct cause is array builder type mismatch, the root cause may be the inconsistent derivation logic between frontend and backend.

**Comment by yufansong** [comment]
Solved at #14398
